openedx/core/djangoapps/content_tagging/api.py

```diff
@@ -175,6 +190,118 @@ def tag_content_object(
     return get_content_tags(object_key, taxonomy_id=taxonomy.id)
 
 
+def get_object_tree_with_objecttags(
+    content_key: ContentKey,
+    include_children: bool = True,
+) -> tuple[TaggedContent, TaxonomyDict]:
+    """
+    Returns the object with the tags associated with it. If include_children is True, then it will also include
+    the children of the object and their tags.
+    """
+
+    def _get_object_tags(content_key: ContentKey, include_children: bool) -> QuerySet[ObjectTag]:
+        """
+        Return the tags for the object and its children using a single db query.
+        """
+        content_key_str = str(content_key)
+        if not include_children:
+            return ObjectTag.objects.filter(
+                object_id=content_key_str,
+                tag__isnull=False,
+                tag__taxonomy__isnull=False,
+            ).select_related("tag__taxonomy")
+
+        # We use a block_id_prefix (i.e. the modified course id) to get the tags for the children of the Content
+        # (course) in a single db query.
+        # ToDo: Add support for other content types (like LibraryContent and LibraryBlock)
+        if isinstance(content_key, UsageKey):
+            course_key_str = str(content_key.course_key)
+            block_id_prefix = course_key_str.replace("course-v1:", "block-v1:", 1)
```

Comment:
This optimization attempted to load all required ObjectTags beforehand with a single query. The other alternatives were:
- run a query for each xblock to get its tags
- create an id list and query the object tags after catching all blocks and iterate over the tree again to associate ObjectTags -> XBlock

This should go because we will remove the UsageKey option.